Sources/ApolloTestSupport/MockGraphQLServer.swift

```diff
@@ -0,0 +1,73 @@
+@testable import Apollo
+import XCTest
+
+public class MockGraphQLServer {
+  enum ServerError: Error, CustomStringConvertible {
+    case unexpectedRequest(String)
+    
+    public var description: String {
+      switch self {
+      case .unexpectedRequest(let requestDescription):
+        return "Mock GraphQL server received an unexpected request: \(requestDescription)"
+      }
+    }
+  }
+  
+  public typealias RequestHandler<Operation: GraphQLOperation> = (HTTPRequest<Operation>) -> JSONObject
+  
+  private class RequestExpectation<Operation: GraphQLOperation>: XCTestExpectation {
+    let file: StaticString
+    let line: UInt
+    let handler: RequestHandler<Operation>
+    
+    init(description: String, file: StaticString = #filePath, line: UInt = #line, handler: @escaping RequestHandler<Operation>) {
+      self.file = file
+      self.line = line
+      self.handler = handler
+
+      super.init(description: description)
+    }
+  }
+  
+  private let queue = DispatchQueue(label: "com.apollographql.MockGraphQLServer")
+    
+  public init() { }
+  
+  // Using existentials to avoid having to deal with type erasure.
+  private var requestExpectations: [AnyHashable: Any] = [:]
+  
+  private subscript<Operation: GraphQLOperation>(_ operationType: Operation.Type) -> RequestExpectation<Operation>? {
+    get {
+      requestExpectations[ObjectIdentifier(operationType)] as! RequestExpectation<Operation>?
+    }
+    
+    set {
+      requestExpectations[ObjectIdentifier(operationType)] = newValue
+    }
+  }
+  
+  public func expect<Operation: GraphQLOperation>(_ operationType: Operation.Type, file: StaticString = #filePath, line: UInt = #line, requestHandler: @escaping (HTTPRequest<Operation>) -> JSONObject) -> XCTestExpectation {
+    return queue.sync {
+      let expectation = RequestExpectation<Operation>(description: "Served request for \(String(describing: operationType))", file: file, line: line, handler: requestHandler)
+      expectation.assertForOverFulfill = true
+      
+      self[operationType] = expectation
+      
+      return expectation
+    }
+  }
+  
+  func serve<Operation>(request: HTTPRequest<Operation>, completionHandler: @escaping (Result<JSONObject, Error>) -> Void) where Operation: GraphQLOperation {
+    // Dispatch after a small random delay to spread out concurrent requests and simulate somewhat real-world conditions.
```

Comment:
Oh, sorry, this was hidden when I replied before. No, it's not really about solving flakiness, but about spreading out concurrent requests over time. If you run concurrent operations within `DispatchQueue.concurrentPerform` (or by manually dispatching to a concurrent queue), many operations will run very close together. And that means the behavior you're testing is pretty artificial. I noticed this in these watcher tests for example, because even if a lot of concurrent requests were running, I would still only get a single update because the fetches were being cancelled before they started. That meant I wasn't seeing any thread sanitizer issues either, and that was what I was trying to reproduce.

Reply:
Right, I'm not saying it's about solving flakiness, I'm talking about when flakiness crops up in the future because some things take long enough that there's not a problem and some things don't

Reply:
Yeah, I realize introducing randomness into tests is a risk, but I'm not sure what the alternative is. We can't use a fixed delay because that would still mean concurrent invocations all complete at the same time. So I do think we need a way to spread request out. Do you have any ideas?

Reply:
Hm, good point about the concurrent invocations all occurring at an identical time...I think I'm fine with this if we can come up with a good story to tell ourselves about how to troubleshoot this when only some invocations are failing. 

You got a good story for that? 🙃

Reply:
No good story once that happens, but if any tests fail because they rely on the difference between a request taking 10ms or 50ms it seems something has gone terribly wrong :)

Reply:
Heh, fair point - we should definitely keep an eye out for problems here though